itest/rust/src/engine_tests/native_audio_structures_test.rs

```diff
@@ -0,0 +1,149 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::thread;
+use std::time::Duration;
+
+use godot::builtin::Vector2;
+use godot::classes::native::AudioFrame;
+use godot::classes::{
+    AudioEffect, AudioEffectInstance, AudioServer, AudioStreamGenerator,
+    AudioStreamGeneratorPlayback, AudioStreamPlayer, Engine, IAudioEffect, IAudioEffectInstance,
+    SceneTree,
+};
+use godot::global::godot_print;
+use godot::obj::{Base, Gd, NewAlloc, NewGd};
+use godot::register::{godot_api, GodotClass};
+
+use crate::framework::itest;
+
+#[derive(GodotClass)]
+#[class(base = AudioEffect, init)]
+struct AudioEffectReceiver {
+    base: Base<AudioEffect>,
+}
+
+#[godot_api]
+impl IAudioEffect for AudioEffectReceiver {
+    fn instantiate(&mut self) -> Option<Gd<AudioEffectInstance>> {
+        Some(AudioEffectReceiverInstance::new_gd().upcast())
+    }
+}
+
+#[derive(GodotClass)]
+#[class(base = AudioEffectInstance, init)]
+struct AudioEffectReceiverInstance {
+    was_called: bool,
+    base: Base<AudioEffectInstance>,
+}
+
+#[godot_api]
+impl IAudioEffectInstance for AudioEffectReceiverInstance {
+    unsafe fn process(
+        &mut self,
+        _src_buffer: *const std::ffi::c_void,
+        dst_buffer: *mut AudioFrame,
+        _frame_count: i32,
+    ) {
+        (*dst_buffer).left = 15.0;
+        (*dst_buffer).right = -12.0;
+        self.was_called = true;
+    }
+}
+
+#[derive(GodotClass)]
+#[class(base = AudioEffect, init)]
+struct AudioEffectAsserter {
+    base: Base<AudioEffect>,
+}
+
+#[godot_api]
+impl IAudioEffect for AudioEffectAsserter {
+    fn instantiate(&mut self) -> Option<Gd<AudioEffectInstance>> {
+        Some(AudioEffectAsserterInstance::new_gd().upcast())
+    }
+}
+
+#[derive(GodotClass)]
+#[class(base = AudioEffectInstance, init)]
+struct AudioEffectAsserterInstance {
+    was_called: bool,
+    base: Base<AudioEffectInstance>,
+}
+
+#[godot_api]
+impl IAudioEffectInstance for AudioEffectAsserterInstance {
+    unsafe fn process(
+        &mut self,
+        src_buffer: *const std::ffi::c_void,
+        _dst_buffer: *mut AudioFrame,
+        _frame_count: i32,
+    ) {
+        let src = src_buffer as *const AudioFrame;
+
+        assert_eq!((*src).left, 15.0);
+        assert_eq!((*src).right, -12.0);
+
+        self.was_called = true;
+    }
+}
+
+#[itest]
+fn native_audio_structure_out_parameter() {
+    // Create two audio effects: One that writes to the out parameter and one that reads the incomming parameter.
+    let receiver = AudioEffectReceiver::new_gd();
+    let asserter = AudioEffectAsserter::new_gd();
+
+    let mut audio_server = AudioServer::singleton();
+
+    // Add both effects to the audio bus so they get called during audio processing.
+    audio_server.add_bus_effect(0, receiver.clone().upcast::<AudioEffect>());
+    audio_server.add_bus_effect(0, asserter.clone().upcast::<AudioEffect>());
```

Comment:
@Bromeon cloning here causes three leaks. Not cloning causes the not-null assertion to fail in `add_bus_effect`.
```
ERROR: Condition "p_effect.is_null()" is true.
   at: add_bus_effect (servers/audio_server.cpp:1002)
```
`AudioEffect` is reference counted so this looks to me like the ref count is off.

Reply:
I was able to track down the cause for the leak:
1. The Gd<AudioEffect> is being moved into `AudioServer::add_bus_effect`.
2. It is then moved it into `AudioServer::add_bus_effect_ex`. 
3. There `as_object_arg()` is called which takes a raw pointer to the object
4. The `ObjectArg` is then moved into a `ExAddBusEffect`. 
5. This struct is then returned from the `add_bus_effect_ex`
6. The `Gd<AudioEffect>` is dropped at the end of`add_bus_effect_ex`
8. This decrements the reference count of the `Gd`.

Reply:
Yes, this was reported in #835 :wink: but it's fixed now.

Cloning shouldn't cause leaks though. Each cloned instance is dropped at some point.

Reply:
> Yes, this was reported in https://github.com/godot-rust/gdext/issues/835 😉 but it's fixed now.

Nice.

> Cloning shouldn't cause leaks though. Each cloned instance is dropped at some point.

Yes, the use-after-free lead me down the wrong path. The leaks do not seem to be related to the clone. Leaks occur when using the `AudioServer` APIs, but I haven't spent any time digging deeper yet.

Reply:
If the audio stuff adds too much headache, it's totally fine for me if you take it out of the PR and we can look at it later. You already did a tremendous job with the virtual required methods.

There have now been how many un- or semi-related issues? :grimacing: 
- dealing with hundreds of LoC
- #842
- #835
- the minimal/full feature
- a leak?

So, don't overdo it :blush: